src/foundation.rs

```diff
@@ -256,3 +258,75 @@ pub trait NSDate {
 impl NSDate for id {
 
 }
+
+#[repr(C)]
+struct NSFastEnumerationState {
+    pub state: libc::c_ulong,
+    pub items_ptr: *mut id,
+    pub mutations_ptr: *mut libc::c_ulong,
+    pub extra: [libc::c_ulong; 5]
+}
+
+pub struct NSFastIterator {
+    state: NSFastEnumerationState,
+    buffer: [id; 16],
+    mut_val: Option<libc::c_ulong>,
+    len: usize,
+    idx: usize,
+    object: id
+}
+
+impl Iterator for NSFastIterator {
+    type Item = id;
+
+    fn next(&mut self) -> Option<id> {
+        if self.idx >= self.len {
+            self.len = unsafe {
+                msg_send![self.object, countByEnumeratingWithState:&mut self.state objects:self.buffer count:16]
+            };
+            self.idx = 0;
+        }
+
+        let new_mut = unsafe {
+            *self.state.mutations_ptr
+        };
+
+        if let Some(old_mut) = self.mut_val {
+            assert!(old_mut == new_mut, "The collection was mutated while being enumerated");
+        }
+
+        if self.idx < self.len {
+            let object = unsafe {
+                let objects: &[id] = slice::from_raw_parts(self.state.items_ptr, self.len);
+                objects[self.idx]
```

Comment:
Up to you, but instead of constructing a slice this could be `*self.state.items_ptr.offset(self.idx)`. Not sure whether constructing a slice or using raw pointer offsets is better when writing unsafe code.